zio: Introduce ft5336 touch sensor driver and integrate into LittlevGL

MaureenHelm opened the pull request:
Depends on #14008 

- Implements the zio interface for the ft5336 capacitive touch sensor on mimxrt10{50,60}_evk boards
- Adds support in LittlevGL to use the zio interface to read touch sensor input data

cc: @bfrog

MaureenHelm committed:
dts: boards: Add ft5336 touch sensor dts bindings and nodes

Adds device tree bindings and nodes for the ft5336 capacitive touch
sensor to the mimxrt10{50,60,64}_evk boards.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
drivers: zio: Introduce ft5336 touch sensor driver

Introduces a new zio driver for the ft5336 capacitive touch sensor.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
lib: gui: Add support for touch input to LittlevGL integration

Adds support for LittlevGL input devices using the new zio interface.
This should ultimately be independent of the specific touch controller,
however it is currently tied to the ft5336 due to the lack of an
abstracted touch sensor data type.

We also have not implemented gpio interrupts in the ft5336 driver and
therefore rely on manual triggering, making the touch input appear a bit
laggy in the gui.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: Configure LittlevGL input defaults on mimxrt10{50,60,64}_evk

Configures LittlevGL to use the ft5336 capacitive touch sensor input by
default on the mimxrt10{50,60,64}_evk boards.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
samples: gui: Hack the LittlevGL sample to use touch input

Enables the touch sensor input and adds a keyboard object to the
LittlevGL sample application.

This is a hack, do not merge!

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

bfrog commented:
First of all, very cool that you made this, this is as far from my use case I was envisioning as I could imagine and yet I think it still works quite well here. I'm interested to know your thoughts on where to proceed next with this interface given this unique and fun use case

ulfalizer commented on lib/gui/lvgl/Kconfig:
```diff
@@ -216,6 +216,15 @@ config LVGL_SCREEN_REFRESH_PERIOD
+
+config LVGL_INPUT_POINTER_DEV_NAME
+	string "Input pointer device name"
+	default "INPUT"
```
Maybe a `depends on LVGL_INPUT_POINTER` would make sense here, so that this symbol doesn't show up in the menu when `LVGL_INPUT_POINTER` is disabled.

microbuilder commented on drivers/zio/ft5336/ft5336.c:
```diff
@@ -0,0 +1,189 @@
+		u8_t yl;
+		u8_t reserved[2];
+	} touch[MAX_TOUCHES];
+} __packed;
```
How do we know which of the 5 possible touch point events are valid? Is it worth having some sort of `.valid` flag to indicate that a particular record is valid, or should we be checking for 0 values to know if `touch[x]` contains useful data (one finger used, three fingers, etc.)?

microbuilder commented on drivers/zio/ft5336/ft5336.c:
```diff
@@ -0,0 +1,189 @@
+static const struct zio_dev_api ft5336_driver_api = {
+	.set_attr = NULL,
+	.get_attr = NULL,
+	.get_attrs = NULL,
```
Some sort of attribute may be useful to have here as a reference, even if it's something simple like exposing the driver version number. In my opinion, exposing a driver version in a standard manner might even be worth considering as a mandatory part of zio since it has little overall cost, but is a useful datapoint when debugging remote devices in the field, or reporting device information.

Alternatively, perhaps the `Operation Mode` of the FT5336 could be exposed as an attribute (Active, Monitor or Hibernate looking at the DS)?

MaureenHelm replied:
Where did you find the DS??? I couldn't find it, so I used the MCUXpresso SDK driver as a reference.

microbuilder replied:
https://www.crystalfontz.com/controllers/FocalTech/FT5336GQQ/

MaureenHelm commented:
Rebased now that #14008 and #16262 have been merged, and added rt1064 support.

teburd reviewed (approved):
This looks like a nice beginning

MaureenHelm committed:
zio: Add support for wrapping/unwrapping u8_t variants

Prior to this patch, zio only supported u32_t variants. Add support for
wrapping and unwrapping u8_t variants so the ft5336 driver can implement
a new u8_t attribute.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
drivers: zio: Add attribute to ft5336 driver to configure touches

Adds a read/write device attribute to the ft5336 zio driver to runtime
configure the number of touches to push to the fifo. The ft5336 hardware
can support up to five simultaneous touches, but some applications like
LittlevGL only use a single touch.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
lib: gui: Configure LittlevGL touch input driver to single touch

LittlevGL doesn't suport multitouch inputs, so use the zio attribute
setter interface to configure the ft5336 driver to only push single
touches to the zio fifo.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm commented:
Rebased to latest topic-sensors branch now that #16266 has been merged. Adds three new commits to implement getting and setting a device attribute.

vanwinkeljan commented:
@MaureenHelm just a heads up but I have an outstanding PR (#17629) to introduce LVGL V6.
As the new version has multiple display support it break API backwards compatibility with V5.

vanwinkeljan commented on samples/gui/lvgl/src/main.c:
```diff
@@ -37,6 +37,32 @@ void main(void)
 	count_label = lv_label_create(lv_scr_act(), NULL);
 	lv_obj_align(count_label, NULL, LV_ALIGN_IN_BOTTOM_MID, 0, 0);
 
+	/*Create styles for the keyboard*/
```
In case it is of any use I have a [sample](https://github.com/vanwinkeljan/zephyr/tree/sdl_gpio/samples/gui/lvgl_keypad) floating around that uses GPIO pins to control some switches maybe this could be extended to allow controle through the touch screen.

carlescufi commented:
@MaureenHelm since `zio` has now been replaced with `rtio`, this PR needs rebasing against the new API.
